tests/mapper/test_adjacency_generators.py

```diff
@@ -0,0 +1,121 @@
+
+    generator = AdjacencyPowerGenerator(barbell, num_powers=5)
+    num_nodes = len(barbell.nodes())
+
+    assert generator.num_powers == 5
+    assert generator.Aadj_T.shape == (num_nodes, num_nodes)
+    assert generator.transition_matrix_T.shape == (num_nodes, num_nodes)
+
+
+def test_bad_init(barbell):
+
+    with pytest.raises(TypeError):
+        generator = AdjacencyPowerGenerator(None, num_powers=5)
+
+    with pytest.raises(ValueError, match="num_powers: expected.*found -1"):
+        generator = AdjacencyPowerGenerator(barbell, num_powers=-1)
+
+    with pytest.raises(TypeError, match="num_powers: expected.*found float"):
+        generator = AdjacencyPowerGenerator(barbell, num_powers=1.0)
+
+
+def test_flow(barbell):
+
+    generator = AdjacencyPowerGenerator(barbell, num_powers=5)
+    for i, x in enumerate(generator.flow(batch_size=1)):
+
+        if i > 5:
+            break
+
+    assert i > 5
+
+
+def test_bad_flow(barbell):
+
+    generator = AdjacencyPowerGenerator(barbell, num_powers=5)
+
+    with pytest.raises(ValueError, match="batch_size: expected.*found 0"):
+        generator.flow(0, num_parallel_calls=1)
+
+    with pytest.raises(TypeError, match="batch_size: expected.*found float"):
+        generator.flow(1.0, num_parallel_calls=1)
+
+    with pytest.raises(ValueError, match="num_parallel_calls: expected.*found 0"):
+        generator.flow(1, num_parallel_calls=0)
+
+    with pytest.raises(TypeError, match="num_parallel_calls: expected.*found float"):
+        generator.flow(1, num_parallel_calls=1.01)
+
+
+@pytest.mark.parametrize("batch_size", [1, 5, 10])
+def test_flow_batch_size(barbell, batch_size):
+
+    num_powers = 5
+    generator = AdjacencyPowerGenerator(barbell, num_powers=num_powers)
+    for x, y in generator.flow(batch_size=batch_size):
+
+        assert x[0].shape == (batch_size,)
+        assert y.shape == (batch_size, 1, len(barbell.nodes()))
+        assert x[1].shape == (batch_size, num_powers, len(barbell.nodes()))
+
+        break
+
+
+@pytest.mark.parametrize("num_powers", [2, 4, 8])
+def test_flow_batch_size(barbell, num_powers):
+
+    batch_size = 2
+    generator = AdjacencyPowerGenerator(barbell, num_powers=num_powers)
+    for x, y in generator.flow(batch_size=batch_size):
+
+        assert x[0].shape == (batch_size,)
+        assert y.shape == (batch_size, 1, len(barbell.nodes()))
+        assert x[1].shape == (batch_size, num_powers, len(barbell.nodes()))
+
+        break
+
+
+@pytest.mark.parametrize("num_powers", [2, 4, 8])
+def test_partial_powers(barbell, num_powers):
+
+    Aadj = normalize_adj(barbell.to_adjacency_matrix(), symmetric=False).todense()
+    actual_powers = [Aadj]
+    for _ in range(num_powers - 1):
+        actual_powers.append(actual_powers[-1].dot(Aadj))
+
+    generator = AdjacencyPowerGenerator(barbell, num_powers=num_powers)
+    for i, (x, y) in enumerate(generator.flow(batch_size=1)):
+
+        if i == len(barbell.nodes()):
+            break
```

Comment:
Why is this break necessary? Doesn't `enumerate` only generate from `0`th to `n-1`th items?

Reply:
The dataset loops infinitely, so this breaks the loop once all the powers have been tested

Reply:
But I can get rid of the `break` by using `generator.flow(batch_size=1).take(len(barbell.nodes()))`